subsys/bluetooth/services/ras/rreq/ras_rreq.c

```diff
@@ -0,0 +1,714 @@
+/*
+ * Copyright (c) 2024 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <zephyr/bluetooth/conn.h>
+#include <zephyr/bluetooth/uuid.h>
+#include <zephyr/bluetooth/gatt.h>
+#include <zephyr/bluetooth/ead.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/sys/byteorder.h>
+#include <zephyr/sys/check.h>
+#include <zephyr/net_buf.h>
+#include <bluetooth/services/ras.h>
+#include <bluetooth/gatt_dm.h>
+#include <stdint.h>
+#include <errno.h>
+
+#include "../ras_internal.h"
+#include "zephyr/sys/__assert.h"
+
+LOG_MODULE_REGISTER(ras_rreq, CONFIG_BT_RAS_RREQ_LOG_LEVEL);
+
+enum bt_ras_rreq_cp_state {
+	BT_RAS_RREQ_CP_STATE_NONE,
+	BT_RAS_RREQ_CP_STATE_GET_RD_WRITTEN,
+	BT_RAS_RREQ_CP_STATE_ACK_RD_WRITTEN,
+};
+
+struct bt_ras_rreq_cp {
+	struct bt_gatt_subscribe_params subscribe_params;
+	enum bt_ras_rreq_cp_state state;
+};
+struct bt_ras_on_demand_rd {
+	struct bt_gatt_subscribe_params subscribe_params;
+	struct net_buf_simple *ranging_data_out;
+	bt_ras_rreq_ranging_data_get_complete_t cb;
+	uint16_t counter_in_progress;
+	uint8_t next_expected_segment_counter;
+	bool data_get_in_progress;
+	bool last_segment_received;
+	bool error_with_data_receive;
+};
+
+struct bt_ras_rd_ready {
+	struct bt_gatt_subscribe_params subscribe_params;
+	bt_ras_rreq_rd_ready_cb_t cb;
+};
+
+struct bt_ras_rd_overwritten {
+	struct bt_gatt_subscribe_params subscribe_params;
+	bt_ras_rreq_rd_overwritten_cb_t cb;
+};
+
+struct bt_ras_rreq {
+	struct bt_conn *conn;
+	struct bt_ras_rreq_cp cp;
+	struct bt_ras_on_demand_rd on_demand_rd;
+	struct bt_ras_rd_ready rd_ready;
+	struct bt_ras_rd_overwritten rd_overwritten;
+};
+
+static struct bt_ras_rreq m_rreq_pool[CONFIG_BT_RAS_RREQ_MAX_ACTIVE_CONN];
+
+static struct bt_ras_rreq *ras_rreq_find(struct bt_conn *conn)
+{
+	for (size_t i = 0; i < ARRAY_SIZE(m_rreq_pool); i++) {
+		struct bt_ras_rreq *rreq = &m_rreq_pool[i];
+
+		if (rreq->conn == conn) {
+			return rreq;
+		}
+	}
+
+	return NULL;
+}
+
+static int ras_rreq_alloc(struct bt_conn *conn)
+{
+	struct bt_ras_rreq *rreq = NULL;
+
+	if (ras_rreq_find(conn) != NULL) {
+		return -EALREADY;
+	}
+
+	for (size_t i = 0; i < ARRAY_SIZE(m_rreq_pool); i++) {
+		if (m_rreq_pool[i].conn == NULL) {
+			rreq = &m_rreq_pool[i];
+			break;
+		}
+	}
+
+	if (rreq == NULL) {
+		return -ENOMEM;
+	}
+
+	LOG_DBG("conn %p new rreq %p", (void *)conn, (void *)rreq);
+
+	memset(rreq, 0, sizeof(struct bt_ras_rreq));
+	rreq->conn = bt_conn_ref(conn);
+
+	return 0;
+}
+
+static uint8_t ranging_data_ready_notify_func(struct bt_conn *conn,
+					      struct bt_gatt_subscribe_params *params,
+					      const void *data, uint16_t length)
+{
+
+	if (length != sizeof(uint16_t)) {
+		LOG_DBG("Ranging Data Ready Indication size error");
+		return BT_GATT_ITER_CONTINUE;
+	}
+
+	struct bt_ras_rreq *rreq = ras_rreq_find(conn);
```

Comment:
`ras_rreq_find` should probably never return NULL but _it could_. So we should check and handle that before trying to dereference it.